### DISCUSSION TITLE
Unique IDs for component instances that work in SSR

### DISCUSSION THREAD

**Discussion by jshimkoski** [nuxt/nuxt#16828] [Questions] [answered]
**Here is the scenario:** A third-party component library needs unique ids for various components to work (think accessibility, form controls, dropdowns, popover, all that fun stuff). Having unique identifiers across component instances is an absolute must.

In Nuxt 3 when using SSR, you cannot do this to assign a unique id due to hydration issues:

```
<script setup>
const id = Math.random()
</script>
```

The server will always assign `id` different to what the client generates later.

In Nuxt 3, there is `useAsyncData` that can be used to make this data the same on the server and on the client. However, the third-party component library is written using Vite and just exposes bundled Vue components via ESM. Nuxt has nothing to do with the library itself.

So, long story short, how could a third-party component library generate unique IDs for component instances that will work in an SSR environment?

Apparently, `@nuxt/kit` can be used to make this situation work in Nuxt 3, but how? Doesn't `@nuxt/kit` only work at the module level? Is there a barebones example of this somewhere?

**Comment by danielroe** [comment]
One thing you could do is pick an ID and encode it in a data attribute, and then read it from your attrs on `created`. (On hydration, this would be present.) If not present you could go ahead and recreate it.

An example of this in Vue 2 can be seen here: https://github.com/nuxt/nuxt.js/blob/dev/packages/vue-app/template/mixins/fetch.client.js#L41

**Reply by jshimkoski** [reply]
Unfortunately, I don't think this solves the need for the id to be automatically generated. It seems to require a hard coded id value pulled in from outside of the component.

If you try to generate that id as required, the same issue occurs, just at a different level of the hierarchy.

Did I misunderstand your suggestion?

**Comment by jshimkoski** [comment]
FYI: I added a feature request to vuejs/core for an `$instanceId` that would solve this.

Likelihood of its acceptance... meh. But, worth a shot anyway: https://github.com/vuejs/core/issues/5468#issue-1146234435

**Comment by danielroe** [comment]
I've released a tiny directive that may help: [vue-bind-once](https://github.com/danielroe/vue-bind-once). Sample usage:

```html
<div v-bind-once="{ id: Math.random() }" />
```

This will work on both server and on client re-hydration.

**Reply by jshimkoski** [reply]
Man, you're fast! I'll check it out today and check back in to see how it goes. 

Thanks.

**Comment by jshimkoski** [comment, marked as answer]
@danielroe solution is pretty awesome, however, it was a bit more robust than what I actually needed.

His work did inspire me to release: [vue-uid](https://github.com/shimyshack/uid)

It generates an SSR-friendly unique identifier that is automatically assigned to the `id` on the element.

Sample usage:

```vue
<script setup lang="ts">
import { ref } from 'vue'
const input = ref<null | HTMLElement>(null)
</script>

<template>
  <div>
    <label :for="input?.id">Input label</label>
    <input v-uid ref="input" type="text">
  </div>
</template>
```

**Reply by webistomin** [reply]
Is there a simple solution for vue 2?

**Reply by webistomin** [reply]
Found a solution, looks a bit ugly but worked for me. You just need to add a `ref` for the problem node

**Reply by simonmaass** [reply]
@webistomin how did you solve this for vue2? @jshimkoski do u have a vue 2 solution?

**Reply by webistomin** [reply]
@simonmaass 

```vue
<template>
  <input :id="id"  ref="forceSsrId" />
</template>

<script>
import { nanoid } from 'nanoid'

export default {
  data () {
    return {
      id: null,
    }
  },

  created () {
    this.id = `id-${nanoid(10)}`
  }
}
</script>
```

**Reply by iliailia1212** [reply]
@webistomin With SSR, the id will be different, this can lead to problems, it is more reliable to check that the id has not yet been set:

```
this.id = this.id || `id-${nanoid(10)}`
```
